### PR TITLE
Use InkProject instead of custom structure for the metadata

### PR DESCRIPTION
**Pull request by obrok**

### CODE REVIEW

File: ink-wrapper/Cargo.toml [hunk 1]
```diff
@@ -12,3 +12,5 @@ clap = { version = "4.1", features = ["derive"] }
 anyhow = "1.0"
 genco = "0.17"
 hex = "0.4.3"
+ink_metadata = "4.0.1"
+scale-info = "2.3"
```
Comment (deuszx) [inline ink-wrapper/Cargo.toml:16]: Can we fix these versions? i.e.
```suggestion
ink_metadata = "=4.0.1"
scale-info = "=2.3"
```
I don't trust anymore that releases will follow SemVer correctly - better safe than sorry.

File: ink-wrapper/src/main.rs [hunk 2]
```diff
@@ -253,177 +177,249 @@ fn generate(metadata: &Metadata, types: &HashMap<u32, &Type>) -> rust::Tokens {
...
-        TypeDef::Primitive { .. } => rust::Tokens::new(),
-        TypeDef::Variant { variant, .. } => define_variant(typ, &variant, types),
-        TypeDef::Composite { composite, .. } => define_composite(typ, &composite, types),
-        TypeDef::Tuple { .. } => rust::Tokens::new(),
+/// Generates a type definition for a custom type used in the contract.
+fn define_type(typ: &Type<PortableForm>, metadata: &InkProject) -> rust::Tokens {
+    match &typ.type_def() {
+        TypeDef::Variant(variant) => define_variant(typ, variant, metadata),
+        TypeDef::Composite(composite) => define_composite(typ, composite, metadata),
+        _ => quote! {},
```
Comment (deuszx) [inline ink-wrapper/src/main.rs:196]: What about [other types](https://docs.rs/scale-info/2.3.1/src/scale_info/ty/mod.rs.html#275-300), like `Array` for example? I tried to use this tool for some `PSP22` contracts IIRC and it failed on this exact case.

**Review by deuszx — COMMENTED** [review]
Without going deep into the project, its dependencies and structure of the contract's spec I won't be able to provide a comprehensive code review that I would like. I guess we will be reacting as we go and hit new cases.

File: ink-wrapper/Cargo.toml (hunk 1 above)
Comment (obrok) [reply ink-wrapper/Cargo.toml:16]: Fair, will do.

File: ink-wrapper/src/main.rs (hunk 2 above)
Comment (obrok) [reply ink-wrapper/src/main.rs:196]: I just wanted to have this PR be the switch and then add more types in another PR.

**Review by deuszx — APPROVED** [review]

**Pull request merged**